Apk to jimple

palaniappan1 opened the pull request.

palaniappan1 committed:
DextoJimple conversion

palaniappan1 committed:
Merge remote-tracking branch 'origin/develop' into ApkToDex

# Conflicts:
#	sootup.core/src/main/java/sootup/core/inputlocation/FileType.java

palaniappan1 committed:
Implemented Trap Handler

palaniappan1 committed:
Changes for branching map

palaniappan1 committed:
Written tests and passing

palaniappan1 committed:
fmt commit

palaniappan1 committed:
Merge remote-tracking branch 'origin/develop' into ApkToDex

palaniappan1 committed:
Added the code flow pdf

palaniappan1 committed:
Merge remote-tracking branch 'origin/develop' into ApkToDex

# Conflicts:
#	sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/PathBasedAnalysisInputLocation.java

palaniappan1 committed:
Merged develop and made the changes

palaniappan1 committed:
Merged with develop and made necessary changes

palaniappan1 committed:
Moved the bodyinterceptors to java.core package

palaniappan1 committed:
Added new instructions handling

palaniappan1 committed:
Added DexTrapStackTransformer

palaniappan1 committed:
Some issue fixes and retype instruction logic

palaniappan1 committed:
InsertBefore Stmt logic added

palaniappan1 committed:
Merge remote-tracking branch 'origin/develop' into ApkToDex

# Conflicts:
#	sootup.core/src/main/java/sootup/core/jimple/basic/LocalGenerator.java
#	sootup.core/src/main/java/sootup/core/model/HasPosition.java
#	sootup.examples/src/test/java/sootup/examples/bodyInterceptor/BodyInterceptor.java
#	sootup.java.bytecode/src/test/java/sootup/java/bytecode/interceptors/CastAndReturnInlinerTest.java
#	sootup.java.bytecode/src/test/java/sootup/java/bytecode/interceptors/LocalSplitterTest.java
#	sootup.java.bytecode/src/test/java/sootup/java/bytecode/interceptors/typeresolving/AugEvalFunctionTest.java
#	sootup.java.bytecode/src/test/java/sootup/java/bytecode/interceptors/typeresolving/TypeResolverTest.java
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/JimpleSemanticsChecker.java
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/LocalSplitter.java

palaniappan1 committed:
Merged with develop

palaniappan1 committed:
Merge remote-tracking branch 'origin/develop' into ApkToDex

# Conflicts:
#	sootup.java.bytecode/src/test/java/sootup/java/bytecode/interceptors/typeresolving/BytecodeHierarchyTest.java
#	sootup.java.bytecode/src/test/java/sootup/java/bytecode/interceptors/typeresolving/TypeAssignerTest.java
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/typeresolving/AugEvalFunction.java
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/typeresolving/BytecodeHierarchy.java
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/typeresolving/PrimitiveHierarchy.java
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/typeresolving/TypePromotionVisitor.java
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/typeresolving/TypeResolver.java
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/typeresolving/types/TopType.java

palaniappan1 committed:
Changed the interceptors to a different package

palaniappan1 committed:
Merge remote-tracking branch 'origin/develop' into ApkToDex

# Conflicts:
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/typeresolving/AugEvalFunction.java
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/typeresolving/BytecodeHierarchy.java
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/typeresolving/PrimitiveHierarchy.java
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/typeresolving/TypePromotionVisitor.java
#	sootup.java.core/src/main/java/sootup/java/core/interceptors/typeresolving/types/TopType.java

palaniappan1 committed:
Merged the interceptor package change and fmt commit

palaniappan1 committed:
Removed unused code to improve codecov

palaniappan1 committed:
Added resolve fields, interfaces, and modifiers

palaniappan1 committed:
Commented the androidversion info check as it is not used yet

palaniappan1 committed:
Resolve Interface done

palaniappan1 committed:
Merge remote-tracking branch 'origin/develop' into ApkToDex

palaniappan1 committed:
fixed 3 instructions code

palaniappan1 committed:
fmt commit

palaniappan1 committed:
Added license headers for 5 files

palaniappan1 committed:
Merge branch 'develop' into ApkToDex

palaniappan1 committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into ApkToDex

palaniappan1 committed:
Updated pom.xml

palaniappan1 committed:
Added one new test, and modified the logic to accommodate new change

swissiety commented on sootup.apk.parser/src/test/java/ApkToDexTest.java:
```diff
@@ -122,6 +114,27 @@ public void testDexClassSource() {
+    ApkAnalysisInputLocation<SootClass> sootClassApkAnalysisInputLocation =
+        new ApkAnalysisInputLocation<>(
+            Paths.get(apk_path),
+            "/Users/palaniappanmuthuraman/Documents/android-platforms",
```
local/absolute path to platform jars

swissiety commented on sootup.apk.parser/src/test/java/ApkToDexTest.java:
```diff
@@ -14,72 +14,64 @@
-  ////        Path path =
-  ////
-  // Paths.get("/Users/palaniappanmuthuraman/Documents/Thesis/Evaluation/Evaluation_TaintBench/apks/playstore_apks");
+  //   As this test uses almost 200 APK's, I have in my local, so this test is not committed.
```
please cleanup ;)

swissiety commented on sootup.apk.parser/src/main/java/ApkAnalysisInputLocation.java:
```diff
@@ -0,0 +1,305 @@
+
+public class ApkAnalysisInputLocation<J extends SootClass> implements AnalysisInputLocation {
+
+  Path apk_path;
```
private final

swissiety commented on sootup.apk.parser/src/main/java/ApkAnalysisInputLocation.java:
```diff
@@ -0,0 +1,305 @@
+    try {
+      dexFromSource =
+          DexFileProvider.getInstance()
+              .getDexFromSource(new File(apk_path.toString()), api_version);
```
apk_path.toFile()?

swissiety commented on sootup.apk.parser/src/main/java/ApkAnalysisInputLocation.java:
```diff
@@ -0,0 +1,305 @@
+  @Nullable
+  @Override
+  public SourceType getSourceType() {
+    return SourceType.Application;
```
in theory we have it configurable everywhere.. but I see no usecase to set it to Library - so i guess we can keep it that way.

swissiety commented on sootup.apk.parser/src/main/java/ApkAnalysisInputLocation.java:
```diff
@@ -0,0 +1,305 @@
+    return bodyInterceptors;
+  }
+
+  //  public String getAndroidJarPath(String jars, String apk) {
```
cleanup; keeping getAndroidAPIVersion() / target sdk() seems reasonable - but refactor that weird getMaxAPIAvailable() stuff into a Utility class as it seems like it useful to automatically determine which runtimes we have - maybe create a subclass of the apkanalysisinputlocation with that code which chooses the best fitting runtime jar as this would segregate the responsibilities - while still beeing able to configure choose manually. -- btw is the runtime jar currently still necessary?

swissiety commented on sootup.apk.parser/src/main/java/ApkAnalysisInputLocation.java:
please add a usage example to the docs/analysisinput.md

swissiety commented on sootup.apk.parser/src/main/java/Util/DexUtil.java:
```diff
@@ -0,0 +1,119 @@
+        type = PrimitiveType.DoubleType.getInstance();
+        break;
+      case 'L': // object
+        if (Util.isByteCodeClassName(typeDescriptor)) {
```
are there situations where the condition is false?

swissiety commented on sootup.apk.parser/src/main/java/Util/DexUtil.java:
```diff
@@ -0,0 +1,119 @@
+  }
+
+  public static JavaClassType stringToJimpleType(String className) {
+    return JavaIdentifierFactory.getInstance().getClassType(DexUtil.toQualifiedName(className));
```
the identifierfactory instance should come from the view

swissiety commented on sootup.apk.parser/src/main/java/Util/DexUtil.java:
```diff
@@ -0,0 +1,119 @@
+        type = VoidType.getInstance();
+        break;
+      case '[': // array
+        Type sootType = toSootType(typeDescriptor, pos + 1);
+        if (sootType != null) {
+          type = Type.createArrayType(sootType, 1);
+        }
```
if this is the only call where we call the second parameter != 0.. lets use a do-while loop and return in the other cases directly without the intermediary type assignment

swissiety replied:
ouh this case can lead to null as well.. if that is copied from the bytecodefrontend please adapt it there as well

swissiety commented on sootup.apk.parser/src/main/java/Util/DexUtil.java:
```diff
@@ -0,0 +1,119 @@
+      }
+    }
+    return modifierEnumSet;
+  }
```
as this is a code duplicate - lets move that method into the *Modifiers

swissiety commented on sootup.apk.parser/src/main/java/Util/Util.java:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  public static boolean isByteCodeClassName(String className) {
+    return ((className.startsWith("L") || className.startsWith("["))
```
charAt(0) + caching?

swissiety commented on sootup.apk.parser/src/main/java/Util/Util.java:
```diff
@@ -0,0 +1,73 @@
+      stmt = Jimple.newReturnVoidStmt(StmtPositionInfo.getNoStmtPositionInfo());
+    }
+    return stmt;
+  }
```
multiple uses? inline?

swissiety commented on sootup.apk.parser/src/main/java/Util/Util.java:
```diff
@@ -0,0 +1,73 @@
+    EnumSet<MethodModifier> modifiers = EnumSet.noneOf(MethodModifier.class);
+
+    return modifiers;
+  }
```
like the other comment for Modifiers - are 2 Util classes necessary?

swissiety commented on sootup.apk.parser/src/main/java/dexpler/DexClassProvider.java:
```diff
@@ -0,0 +1,99 @@
+  @Override
+  public Optional<SootClassSource> createClassSource(
+      AnalysisInputLocation inputLocation, Path sourcePath, ClassType classSignature) {
+    // TODO : In soot, they wrote this. This code makes sense, but dont know where to add this or
```
it seems in soot it was a mechanism to make the input apk available to the sourcelocator of soot - we dont have that anymore - this is covered via the analysisinputlocation in sootup

swissiety commented on sootup.apk.parser/src/main/java/dexpler/DexClassSource.java:
```diff
@@ -0,0 +1,197 @@
+        return Optional.ofNullable(DexUtil.stringToJimpleType(superclass));
+      }
+    }
+    return Optional.empty();
```
when does classInformation == null happen? maybe we should throw throw an exception?

swissiety commented on sootup.apk.parser/src/main/java/dexpler/DexFileProvider.java:
```diff
@@ -0,0 +1,220 @@
+    return instance;
+  }
+
+  private static final Comparator<DexContainer<? extends DexFile>> DEFAULT_PRIORITIZER =
```
lets move that comparator to the end of the file

swissiety commented on sootup.apk.parser/src/main/java/dexpler/DexLibWrapper.java:
```diff
@@ -0,0 +1,109 @@
+   */
+  public DexLibWrapper(File dexSource) {
+    try {
+      // TODO Change the api_version to some common place
```
TODO: make it configurable starting from the AnalysisInputLocation

swissiety commented on sootup.apk.parser/src/main/java/dexpler/DexLibWrapper.java:
```diff
@@ -0,0 +1,109 @@
+          // systemAnnotationNames.contains(sootTypeName)) {
+          //                            // dex files contain references to the Type IDs of void
+          //                            // primitive types - we obviously do not want them
+          //                            // to be resolved
```
are we handling that? do we have a testcase that checks for it?

swissiety commented on sootup.apk.parser/src/main/java/dexpler/DexMethod.java:
```diff
@@ -0,0 +1,58 @@
+    int modifierFlags = method.getAccessFlags();
+    if (Modifier.isAbstract(modifierFlags) || Modifier.isNative(modifierFlags)) {
+      String className = declaringclassType.getClassName();
+      if (Util.Util.isByteCodeClassName(className)) {
```
see question above

swissiety commented on sootup.apk.parser/src/main/java/dexpler/DexMethodSource.java:
```diff
@@ -0,0 +1,116 @@
+
+  @Override
+  public Object resolveAnnotationsDefaultValue() {
+    return null;
```
throw new UnsupportedOperation("TODO") exception

swissiety commented on sootup.apk.parser/src/main/java/dexpler/DexMethodSource.java:
```diff
@@ -0,0 +1,116 @@
+      }
+    }
+    return modifierEnumSet;
+  }
```
duplicate?

swissiety commented on sootup.apk.parser/src/main/java/dexpler/DexMethodSource.java:
```diff
@@ -0,0 +1,116 @@
+  @Nonnull
+  @Override
+  public MethodSignature getSignature() {
+    //    String className = classType.getClassName();
```
cleanup

swissiety commented on sootup.apk.parser/src/main/java/dexpler/DexResolver.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.Map;
+import java.util.TreeMap;
+
+public class DexResolver {
```
do we need that cache? i.e. it seems like we are giving a dexlibwrapper to the converted class?
is the given wrapper (and its memory) released at some point?

palaniappan1 replied:
To retrieve each and every class, we need to look into its details from the given dex file. The cache over there read the dex file only once and store all the contents of the class. So every time we request for the details of the class, lookup from the cache and return the information. In short, caching over there is necessary.

swissiety commented on sootup.apk.parser/src/main/java/instruction/AgetInstruction.java:
```diff
@@ -0,0 +1,39 @@
+    Local l = body.getRegisterLocal(dest);
+
+    JAssignStmt assignStmt =
+        Jimple.newAssignStmt(l, jArrayRef, StmtPositionInfo.getNoStmtPositionInfo());
```
do we have access to line numbers?

swissiety commented on sootup.apk.parser/src/main/java/instruction/AputInstruction.java:
```diff
@@ -0,0 +1,54 @@
+    Local sourceValue = body.getRegisterLocal(source);
+    JAssignStmt assign = getAssignStmt(body, sourceValue, jArrayRef);
+    if (aPutInstr.getOpcode() == Opcode.APUT_OBJECT) {
+      // TODO : Need to add one field (tag) with both getters and setters in the statement class
```
.

palaniappan1 replied:
Not needed anymore.

swissiety commented on sootup.apk.parser/src/main/java/instruction/ArrayLengthInstruction.java:
```diff
@@ -0,0 +1,36 @@
+    JAssignStmt assignStmt =
+        Jimple.newAssignStmt(
+            body.getRegisterLocal(dest), jLengthExpr, StmtPositionInfo.getNoStmtPositionInfo());
+    setStmt(assignStmt);
+    body.add(assignStmt);
```
why/necessity of setStmt and add?

palaniappan1 replied:
setStmt() is to assign the jimple equivalent of the dex/bytecode statement. So I have a map to maintain for every given dex instruction its corresponding Jimple Statement. It helps me in easy lookup later (when doing dangling instructions or branching statements)

addStmt() is to assign to the method body.

In my defense, both are needed I think :)

swissiety replied:
seems reasonable. Just as a suggestion to minimize the necessary code: you could iterate over your map of instr->stmt at the end of conversion and add then add the stmts to the body (not sure about the markerunit replacement implementation but replacing them in the body - which is a list by then - so has O(n) to find the item would not be necessary )

swissiety commented on sootup.apk.parser/src/main/java/instruction/CastInstruction.java:
```diff
@@ -0,0 +1,95 @@
+        return PrimitiveType.DoubleType.getInstance();
+
+      default:
+        throw new RuntimeException("Invalid Opcode: " + opcode);
```
IllegalState

swissiety commented on sootup.apk.parser/src/main/java/instruction/BinopLitInstruction.java:
```diff
@@ -0,0 +1,114 @@
+        return Jimple.newUshrExpr(source1, source2);
+
+      default:
+        throw new RuntimeException("Invalid Opcode: " + opcode);
```
IllegalState

swissiety commented on sootup.apk.parser/src/main/java/instruction/CmpInstruction.java:
```diff
@@ -0,0 +1,78 @@
+    // Type type = null
+    Opcode opcode = instruction.getOpcode();
+    Expr cmpExpr = null;
+    Type type = null;
```
type seems to be unused? can we make use of the type at the local definition? otherwise lets comment the assignments for possible later use

swissiety commented on sootup.apk.parser/src/main/java/instruction/ConditionalJumpInstruction.java:
```diff
@@ -0,0 +1,88 @@
+      body.add(s);
+      setStmt(s);
+    } else {
+      // set marker unit to swap real gotostmt with otherwise
+      body.addDeferredJimplification(this);
+      markerUnit = Jimple.newNopStmt(StmtPositionInfo.getNoStmtPositionInfo());
+      setStmt(markerUnit);
```
still necessary? wouldnt it be enought to store the target address to a map and add the branchtarget after everything is resolved?

palaniappan1 replied:
This is one such way of handling stuff. Old School technique :)

swissiety commented on sootup.apk.parser/src/main/java/instruction/ConstStringInstruction.java:
```diff
@@ -0,0 +1,42 @@
+          "Expected Instruction21c or Instruction31c but got neither.");
+    }
+    StringConstant stringConstant =
+        new StringConstant(s, JavaIdentifierFactory.getInstance().getType("java.lang.String"));
```
can we cache the jlString classType? (and getting the identifierfactory from the view)

swissiety commented on sootup.apk.parser/src/main/java/instruction/DexLibAbstractInstruction.java:
```diff
@@ -0,0 +1,111 @@
+   * @param instruction a range invocation instruction
+   * @return a list of register indices
+   */
+  protected List<Integer> getUsedRegistersNums(RegisterRangeInstruction instruction) {
```
depending on the usecase - does a Set make sense here?
lets precalculate the amount of elements in the returned collection

swissiety commented on sootup.apk.parser/src/main/java/instruction/DexLibAbstractInstruction.java:
```diff
@@ -0,0 +1,111 @@
+      instruction.getRegisterF(),
+      instruction.getRegisterG(),
+    };
+    List<Integer> l = new ArrayList<Integer>();
```
lets give asize hint as well

swissiety commented on sootup.apk.parser/src/main/java/instruction/FieldInstruction.java:
```diff
@@ -0,0 +1,106 @@
+      JStaticFieldRef jStaticFieldRef = new JStaticFieldRef(fieldSignature);
+      return jStaticFieldRef;
+    } else {
+      // TODO : Dont know which local to use here, as of now using null which will throw an error
```
.

swissiety commented on sootup.apk.parser/src/main/java/instruction/FillArrayDataInstruction.java:
```diff
@@ -0,0 +1,193 @@
+      }
+    }
+    if (firstAssign == null) { // if numElements == 0. Is it possible?
+      firstAssign = Jimple.newNopStmt(StmtPositionInfo.getNoStmtPositionInfo());
```
lets investigate into that; or defer that question - so just log it or put an assertion and wait until someone complains ;-)

swissiety commented on sootup.apk.parser/src/main/java/instruction/FillArrayDataInstruction.java:
```diff
@@ -0,0 +1,193 @@
+    Instruction referenceTable = body.instructionAtAddress(targetAddress).instruction;
+
+    if (!(referenceTable instanceof ArrayPayload)) {
+      throw new RuntimeException(
```
IllegalState

swissiety commented on sootup.apk.parser/src/main/java/instruction/FillArrayDataInstruction.java:
```diff
@@ -0,0 +1,193 @@
+      }
+    }
+    if (firstAssign == null) { // if numElements == 0. Is it possible?
+      firstAssign = Jimple.newNopStmt(StmtPositionInfo.getNoStmtPositionInfo());
```
same as above

swissiety commented on sootup.apk.parser/src/main/java/instruction/FillArrayDataInstruction.java:
```diff
@@ -0,0 +1,193 @@
+        if (newRegister != -1) {
+          usedRegisters.add(newRegister);
+          usedRegisters.remove(reg);
+          break; // there can't be more than one new
```
do we know something about the structure of the register values except from int? maybe a bitset is more efficient

swissiety commented on sootup.apk.parser/src/main/java/instruction/FillArrayDataInstruction.java:
```diff
@@ -0,0 +1,193 @@
+      value = IntConstant.getInstance(element.intValue());
+      IntConstant ic = (IntConstant) value;
+      if (ic.getValue() != 0) {
+        value = IntConstant.getInstance(1);
```
BooleanConstant?

swissiety commented on sootup.apk.parser/src/main/java/instruction/FillArrayDataInstruction.java:
```diff
@@ -0,0 +1,193 @@
+      value = LongConstant.getInstance(element.longValue());
+    } else {
+      throw new RuntimeException(
+          "Invalid Array Type occured in FillArrayDataInstruction: " + elementType);
```
illegalstate

swissiety commented on sootup.apk.parser/src/main/java/instruction/GotoInstruction.java:
```diff
@@ -0,0 +1,39 @@
+      body.add(jGotoStmt);
+      return;
+    }
+    body.addDeferredJimplification(this);
+    markerUnit = Jimple.newNopStmt(StmtPositionInfo.getNoStmtPositionInfo());
+    setStmt(markerUnit);
```
as above

swissiety commented on sootup.apk.parser/src/main/java/instruction/MethodInvocationInstruction.java:
```diff
@@ -0,0 +1,184 @@
+  protected void jimplifyVirtual(DexBody body) {
+    MethodReference item = (MethodReference) ((ReferenceInstruction) instruction).getReference();
+    List<Local> parameters = buildParameters(body, item.getParameterTypes(), false);
+    // TODO check isIntertface by someother way
```
.

swissiety commented on sootup.apk.parser/src/main/java/instruction/MethodInvocationInstruction.java:
```diff
@@ -0,0 +1,184 @@
+      parameters.add(body.getRegisterLocal(regs.get(i)));
+      // if method is non-static the first parameter is the instance
+      // pointer and has no corresponding parameter type
+      if (!isStatic && i == 0) {
+        j--;
+        continue;
+      }
```
move this check out of the loop

swissiety commented on sootup.apk.parser/src/main/java/instruction/MethodInvocationInstruction.java:
```diff
@@ -0,0 +1,184 @@
+  }
+
+  public static boolean isWide(String type) {
+    return type.startsWith("J") || type.startsWith("D");
```
charAt(0) + reuse

swissiety commented on sootup.apk.parser/src/main/java/instruction/MoveExceptionInstruction.java:
```diff
@@ -0,0 +1,57 @@
+  public void retype(List<Stmt> stmtList, Set<Local> locals) {
+    if (realType == null) {
+      throw new RuntimeException(
+          "Real type of this instruction has not been set or was already retyped: " + this);
```
IllegalState

swissiety commented on sootup.apk.parser/src/main/java/instruction/NopInstruction.java:
```diff
@@ -0,0 +1,22 @@
+  public void jimplify(DexBody body) {
+    JNopStmt jNopStmt = Jimple.newNopStmt(StmtPositionInfo.getNoStmtPositionInfo());
+    setStmt(jNopStmt);
+    // TODO: In one example, there comes NoOp Instruction after the return statement, dont know how
```
.

palaniappan1 replied:
Handled.

swissiety commented on sootup.apk.parser/src/main/java/instruction/PackedSwitchInstruction.java:
```diff
@@ -0,0 +1,29 @@
+    JSwitchStmt jSwitchStmt =
+        Jimple.newLookupSwitchStmt(key, lookupValues, StmtPositionInfo.getNoStmtPositionInfo());
+    //        body.addBranchingStmt(jSwitchStmt, targets);
+    //        body.addBranchingStmt(jSwitchStmt, Collections.singletonList(defaultTarget));
```
.

palaniappan1 replied:
Handled when constructing branching map

swissiety commented on sootup.apk.parser/src/main/java/instruction/PseudoInstruction.java:
```diff
@@ -0,0 +1,71 @@
+
+  public int getDataFirstByte() {
+    if (dataFirstByte == -1) {
+      throw new RuntimeException("Error: dataFirstByte was not set!");
```
IllegalState ; more in this file

swissiety commented on sootup.apk.parser/src/main/java/instruction/ReTypeableInstruction.java:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @author Michael Markert <michael.markert@googlemail.com>
+ */
+public interface ReTypeableInstruction {
```
is this a possible solution to the commented/currenlty unused type assigned variable?

swissiety commented on sootup.apk.parser/src/main/java/instruction/SparseSwitchInstruction.java:
```diff
@@ -0,0 +1,31 @@
+    setStmt(switchStmt);
+    // It is unlike the PackedSwitchInstruction, here only one branching statement should be. (I
+    // literally don't know why :( )
+    //        targets.add(defaultTarget);
```
.

swissiety commented on sootup.apk.parser/src/main/java/instruction/SwitchInstruction.java:
```diff
@@ -0,0 +1,106 @@
+
+  @Override
+  public void jimplify(DexBody body) {
+    markerUnit = Jimple.newNopStmt(StmtPositionInfo.getNoStmtPositionInfo());
```
.

swissiety commented on sootup.apk.parser/src/main/java/main/AndroidVersionInfo.java:
do we still need it? is it part of the commented code or a duplicate in the inputlocation?

swissiety commented on sootup.apk.parser/src/main/java/main/TaggedInstruction.java:
```diff
@@ -0,0 +1,13 @@
+import instruction.DexLibAbstractInstruction;
+import org.jf.dexlib2.iface.instruction.Instruction;
+
+public abstract class TaggedInstruction extends DexLibAbstractInstruction {
```
we removed tags in sootup.. do we use the tags somewhere?

swissiety commented on sootup.core/src/main/java/sootup/core/graph/MutableBlockStmtGraph.java:
please revert

swissiety commented on sootup.core/src/main/java/sootup/core/model/Body.java:
please revert

swissiety commented on sootup.java.core/src/main/java/sootup/java/core/interceptors/Dex/DexDefUseAnalysis.java:
```diff
@@ -0,0 +1,190 @@
+import sootup.core.model.Body;
+
+/* Flow -insensitive def-use analysis */
+public class DexDefUseAnalysis {
```
is it dex specific or would it make sense to move it to the analysis module?

swissiety commented on sootup.java.core/src/main/java/sootup/java/core/interceptors/Dex/DexNumberTranformer.java:
```diff
@@ -0,0 +1,257 @@
+                {
+                  Value rightOp = stmt.getRightOp();
+                  if (rightOp instanceof AbstractBinopExpr && !(rightOp instanceof JCmpExpr)) {
+                    //                        usedAsFloatingPoint = examineBinopExpr(stmt);
```
block is empty?

palaniappan1 replied:
Unwanted check, so removed it now

swissiety commented on sootup.java.core/src/main/java/sootup/java/core/interceptors/Dex/DexTransformer.java:
```diff
@@ -0,0 +1,249 @@
+        JAssignStmt stmt = (JAssignStmt) arrayStmt;
+        aBase = (Local) stmt.getRightOp();
+      } else {
+        throw new RuntimeException("ERROR: not an assign statement: " + arrayStmt);
```
IllegalState

swissiety commented on sootup.java.core/src/main/java/sootup/java/core/interceptors/Dex/DexTransformer.java:
```diff
@@ -0,0 +1,249 @@
+      if (baseDef instanceof JAssignStmt) {
+        JAssignStmt stmt = (JAssignStmt) baseDef;
+        Value r = stmt.getRightOp();
+        if (r instanceof JFieldRef) {
```
ValueVisitor would make sense here to replace the ifelse tree

swissiety reviewed (changes requested):
thanks for your big contribution :) I have a little feedback for you..

palaniappan1 committed:
AndroidVersionInfo moved to a new class

palaniappan1 committed:
Merge remote-tracking branch 'origin/develop' into ApkToDex

palaniappan1 committed:
Test fixed

palaniappan1 committed:
Resolved comments

palaniappan1 committed:
Resolved Comments_1

palaniappan1 committed:
fmt commit

palaniappan1 committed:
Resolved Comments

1. Removed Tagged Instruction
2. Changed RTE to ISException
3. Resolved some comments

kadirayk commented on sootup.apk.parser/resources/stats_file.csv:
why do we need this?

kadirayk commented on sootup.apk.parser/src/main/java/DexBodyInterceptors.java:
are the body interceptors okay so? Or will you decide which ones to activate later?

palaniappan1 committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into ApkToDex

# Conflicts:
#	pom.xml

palaniappan1 committed:
Resolved some comments

palaniappan1 committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into ApkToDex

palaniappan1 committed:
Some minor changes

palaniappan1 committed:
Removed Retypeable instruction and its related code

kadirayk committed:
Merge branch 'develop' into ApkToDex

palaniappan1 committed:
Replaced IOUtils with ByteStreams

palaniappan1 committed:
Merge branch 'ApkToDex' of https://github.com/soot-oss/SootUp into ApkToDex

palaniappan1 committed:
Test issue solved

swissiety commented on sootup.java.core/src/main/java/sootup/java/core/interceptors/Dex/DexNullTransformer.java:
```diff
@@ -372,8 +372,8 @@ public void caseReturnStmt(@Nonnull JReturnStmt stmt) {
-      if (stmt.containsInvokeExpr()) {
-        AbstractInvokeExpr invExpr = stmt.getInvokeExpr();
+      if (stmt.isInvokableStmt() && stmt.asInvokableStmt().containsInvokeExpr()) {
+        AbstractInvokeExpr invExpr = stmt.asInvokableStmt().getInvokeExpr().get();
```
please cache: `stmt.asInvokableStmt()`

swissiety commented on sootup.apk.parser/src/main/java/ApkAnalysisInputLocation.java:
```diff
@@ -1,61 +1,53 @@
 import sootup.core.util.StreamUtils;
 import sootup.core.views.View;
 
 public class ApkAnalysisInputLocation<J extends SootClass> implements AnalysisInputLocation {
```
generics have been removed from AnalysisInputLocation in the meantime

palaniappan1 committed:
Resolved comments
